text/2023-04-26-color-outlines.md

```diff
@@ -0,0 +1,198 @@
+based callback interface is cumbersome.
+
+The proposed API would flatten the graph structure into a linear sequence of
+drawing commands, making use of stack semantics to capture the graph structure
+where necessary. The set of commands required to represent the graph is fairly
+small and maps nicely to most graphics APIs which already represent layering
+with a state stack.
+
+## Interface
+
+This is a rough outline of the proposed interface which would introduce a new
+`ColorPen` trait for accessing color outlines.
+
+### Supporting types
+
+Basic types that are used by the new `ColorPen` trait.
+
+```rust
+/// Extents of a rectangular region.
+pub struct BoundingBox;
+
+/// Affine transformation matrix.
+pub struct Transform;
+
+/// 32-bit RGBA unpremultiplied color.
+pub struct Color;
+
+/// Enumeration containing data for solid or gradient fills.
+pub enum Brush<'a>;
+
+/// Contains a glyph path for clips or fills. Offers a method that can feed
+/// path commands to a Pen.
+pub struct Path<'a>;
+```
+
+### Scaler methods
+
+The `Scaler` type will gain two new methods:
+
+```rust
+fn has_color_outlines(&self) -> bool
+```
+Simply returns true if a `COLR` table is present.
+
+```rust
+fn color_outline(
+    &self, 
+    glyph_id: GlyphId, 
+    palette_fn: impl Fn(u16) -> Option<Color>, 
+    pen: &mut impl ColorPen
+) -> Result<(), Error>
+```
+Loads a color outline for the specified glyph and invokes the appropriate
+callbacks on the pen.
+
+The `palette_fn` parameter is used to specify the mapping from palette
+color index to a color value. This can be used to implement overrides and
+a helper will also be provided for constructing this function for basic
+palette support.
+
+### The `ColorPen` trait
+
+This trait contains methods that are implemented by the caller to access
+the sequence of operations necessary to render a color outline.
+
+> Note: When provided below, the code in block quotes demonstrates how these
+operations might roughly be implemented for a Skia based renderer (assuming
+a local variable `canvas` of type `SkCanvas`)
+
+#### Methods
+
+```rust
+fn bounds(&mut self, bounds: BoundingBox)
+```
+This is guaranteed to be the first method called on the pen to allow an
+appropriately sized surface to be allocated if necessary. It will pass the
+value of the `ClipBox` if available, otherwise a bounding box computed
+during evaluation of the paint graph.
+
+In accordance with the spec, if the paint graph is determined to be
+unbounded, no methods on the pen will be called and the `color_outline`
+method will return an error.
+
+```rust
+fn push_transform(&mut self, transform: Transform)
+```
+Push a transform onto the state stack.
+
+> `canvas->save();`<br>
+`canvas->concat(TransformToSkMatrix(transform));`
+
+```rust
+fn pop_transform(&mut self)
+```
+Pop a transform from the state stack.
+
+> `canvas->restore();`<br>
+
+```rust
+fn push_clip(&mut self, glyph_id: Glyph_id, path: &Path)
```

Comment:
I missed the point about hinting. Skia doesn’t currently hint these outlines. If that requirement changes, we can also handle that internally using the hinting mode that is set when configuring the scaler.

Reply:
In Fontations my impression is we generally follow an approach of providing independent, functional building blocks. I suggest we try to do the same here and decouple the color glyph processing and drawing from other independent parts at useful interfaces. 

I am in favor of creating a separate `GlyphPainter` (name TBD) interface, that's independent of `Scaler` operations.  Where the context of the Scaler is useful for outline caching, the client can use it and pass it in.

The `GlyphPainter` for example, regarding the need for glyph outline caching, the `GlyphPainter` can provide a `bounds()` method that accepts a `Scaler` as an argument, as it needs glyph contour access with glyph rasterisation settings configured.  So, I'd suggest to remove the bounds callback from the `ColorPen` trait . Not only for the reason of decoupling this from the `Scaler` but because I think the bounds method does not really semantically belong in the `ColorPen` as that's not a typical Canvas-like operation.

> First, while inside the `ColorPen` callbacks, the `Scaler` object (and it’s dependencies) have been mutably borrowed and Rust requires that access to be exclusive so you can’t effectively use it to load an outline without resorting to unsafe code.

See above, are there obstacles to decoupling this from Scaler?

Reply:
With this decoupled, we can remove the `path:` argument and have the client use `Scaler` for the purpose.

Reply:
> In Fontations my impression is we generally follow an approach of providing independent, functional building blocks. I suggest we try to do the same here and decouple the color glyph processing and drawing from other independent parts at useful interfaces.

I think there is a slightly blurry line when discussing fontations— read-fonts is intended to provide low level building blocks but skrifa is meant to be a more complete solution. That said, I agree with decoupling this in some way. Finding the correct boundary is a bit of a challenge while taking constraints into account. 

> If outlines are needed, they can be cached in a passed-in `Scaler`'s context, but for the purpose of bounding box calculation in the simple TrueType case, contour retrieval is not needed, is it?

In addition to CFF, variable fonts also need to load outlines for this.

Reply:
> If outlines are needed, they can be cached in a passed-in `Scaler`'s context, but for the purpose of bounding box calculation in the simple TrueType case, contour retrieval is not needed, is it? For the bounding box, then union of each glyphs xMin, xMax, yMin, yMax should be sufficient, which does not require path retrieval? - May be different for COLRv0/v1 + CFF.

Depends on whether you want tight or loose bounds for transformed glyphs.

Reply:
Ah, this is a good point. We didn’t consider transforms in the graph.